src/main/java/io/cdap/plugin/gcp/bigquery/actions/BigQueryMergeAction.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.cdap.plugin.gcp.bigquery.actions;
+
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.BigQueryOptions;
+import com.google.cloud.bigquery.Field;
+import com.google.cloud.bigquery.FieldList;
+import com.google.cloud.bigquery.Schema;
+import com.google.cloud.bigquery.TableId;
+import io.cdap.cdap.api.annotation.Description;
+import io.cdap.cdap.api.annotation.Name;
+import io.cdap.cdap.api.annotation.Plugin;
+import io.cdap.cdap.etl.api.action.Action;
+import io.cdap.cdap.etl.api.action.ActionContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * A action plugin to merge two Big Query tables.
+ */
+@Plugin(type = Action.PLUGIN_TYPE)
+@Name(BigQueryMergeAction.NAME)
+@Description("This action merges two BigQuery tables.")
+public class BigQueryMergeAction extends BigQueryQueryExecute {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryMergeAction.class);
+  public static final String NAME = "BigQueryMerge";
+  private static final String MERGE_QUERY = "MERGE %s T USING %s S ON %s WHEN MATCHED THEN UPDATE SET %s " +
+    "WHEN NOT MATCHED THEN INSERT (%s) VALUES(%s)";
+
+  private BigQueryMergeActionConfig config;
+
+  @Override
+  public void run(ActionContext context) throws Exception {
+    config.validate();
+    LOG.debug("Run merge action.");
+    if (config.isAllowSchemaRelaxation()) {
+      updateTableSchema();
+    }
+    String query = generateQuery();
+    setQuery(query);
+    super.run(context);
+    LOG.debug("Complete merge action.");
+  }
+
+  private String generateQuery() {
```

Comment:
this plugin assumes that the source and destination tables have the same columns, which will not always be true, and which probably is not true in most situations. This plugin needs to go through a design before we implement it. I would imagine the user would need to provide a more generic condition to merge on and select fields from the source need to be updated/inserted in the destination (kind of similar to what is given in the joiner)

@nitinmotgi can you provide more information about the use case? Is it only merging on equality? Do we need this now that there is 'upsert' in the BQ sink and given that there will soon be a generic BigQuery action?